src/dex/generic-rfq/generic-rfq.ts

```diff
@@ -0,0 +1,217 @@
+    for (const [i, amount] of amounts.entries()) {
+      if (amount.isZero()) {
+        outputs[i] = BN_0;
+      } else {
+        let srcAmountLeft = amount.minus(lastTotalSrcAmount);
+        let destAmountFilled = lastTotalDestAmount;
+        while (lastOrderIndex < amountsWithRates.length) {
+          const { amount, price } = amountsWithRates[lastOrderIndex];
+          if (srcAmountLeft.gt(amount)) {
+            const destAmount = amount.multipliedBy(price);
+
+            srcAmountLeft = srcAmountLeft.minus(amount);
+            destAmountFilled = destAmountFilled.plus(destAmount);
+
+            lastTotalSrcAmount = lastTotalSrcAmount.plus(amount);
+            lastTotalDestAmount = lastTotalDestAmount.plus(destAmount);
+            lastOrderIndex++;
+          } else {
+            destAmountFilled = destAmountFilled.plus(
+              srcAmountLeft.multipliedBy(price),
+            );
+            srcAmountLeft = BN_0;
+            break;
+          }
+        }
+        if (srcAmountLeft.isZero()) {
+          outputs[i] = destAmountFilled;
+        } else {
+          // If current amount was unfillable, then bigger amounts are unfillable as well
+          break;
+        }
+      }
+    }
+
+    return {
+      outputs: outputs.map(o =>
+        BigInt(o.multipliedBy(outMultiplier).toFixed(0)),
+      ),
+      ordersCount: lastOrderIndex,
+    };
+  }
+
+  async getPricesVolume(
+    srcToken: Token,
+    destToken: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<ExchangePrices<ParaSwapLimitOrdersData> | null> {
+    const _srcToken = this.dexHelper.config.wrapETH(srcToken);
+    const _destToken = this.dexHelper.config.wrapETH(destToken);
+
+    const _srcAddress = _srcToken.address.toLowerCase();
+    const _destAddress = _destToken.address.toLowerCase();
+    if (_srcAddress === _destAddress) return null;
+
+    const expectedIdentifier = this.getIdentifier(_srcAddress, _destAddress);
+
+    const orderPricesInfo = await this.rateFetcher.getOrderPrice(
+      _srcToken,
+      _destToken,
+      side,
+    );
+    if (!orderPricesInfo) {
+      return null;
+    }
+
+    const inDecimals =
+      side === SwapSide.SELL ? _srcToken.decimals : _destToken.decimals;
+    const outDecimals =
+      side === SwapSide.SELL ? _destToken.decimals : _srcToken.decimals;
+
+    const _amountsInBN = amounts.map(a =>
+      new BigNumber(a.toString()).dividedBy(getBigNumberPow(inDecimals)),
+    );
+
+    const unitVolume = BN_1;
+
+    const unitResults = this.calcOutsFromAmounts(
+      [unitVolume],
+      getBigNumberPow(outDecimals),
+      orderPricesInfo.rates,
+    );
+
+    const unit = unitResults.outputs[0];
+
+    const outputs = this.calcOutsFromAmounts(
+      _amountsInBN,
+      getBigNumberPow(outDecimals),
+      orderPricesInfo.rates,
+    );
+    return [
+      {
+        gasCost: Number(BigInt(outputs.ordersCount) * ONE_ORDER_GASCOST),
```

Comment:
There seems to be some confusion here, the order book is only used to define the price curve, there is no gas cost associated with how many entries we went through.  The firm rate is always returned as one order from the MM.